r/src/array__to_vector.cpp

```diff
@@ -0,0 +1,827 @@
+    return Converter_Time<value_type>::Ingest(data_, array, start, n);
+  }
+};
+
+struct Converter_Int64 {
+  static SEXP Allocate(R_xlen_t n, const ArrayVector&) {
+    NumericVector data(no_init(n));
+    data.attr("class") = "integer64";
+    return data;
+  }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    NumericVector_ data(data_);
+    auto null_count = array->null_count();
+    if (null_count == n) {
+      std::fill_n(reinterpret_cast<int64_t*>(data.begin()) + start, n, NA_INT64);
+    } else {
+      auto p_values = array->data()->GetValues<int64_t>(1);
+      if (!p_values) {
+        return Status::Invalid("Invalid data buffer");
+      }
+
+      auto p_vec = reinterpret_cast<int64_t*>(data.begin()) + start;
+
+      if (array->null_count()) {
+        internal::BitmapReader bitmap_reader(array->null_bitmap()->data(),
+                                             array->offset(), n);
+        for (size_t i = 0; i < n; i++, bitmap_reader.Next()) {
+          p_vec[i] = bitmap_reader.IsNotSet() ? NA_INT64 : p_values[i];
+        }
+      } else {
+        std::copy_n(p_values, n, p_vec);
+      }
+    }
+    return Status::OK();
+  }
+};
+
+struct Converter_Decimal {
+  static SEXP Allocate(R_xlen_t n, const ArrayVector&) {
+    return NumericVector_(no_init(n));
+  }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    NumericVector_ data(data_);
+    auto null_count = array->null_count();
+    if (n == null_count) {
+      std::fill_n(data.begin() + start, n, NA_REAL);
+    } else {
+      auto p_vec = reinterpret_cast<double*>(data.begin()) + start;
+      const auto& decimals_arr =
+          internal::checked_cast<const arrow::Decimal128Array&>(*array);
+
+      if (array->null_count()) {
+        internal::BitmapReader bitmap_reader(array->null_bitmap()->data(),
+                                             array->offset(), n);
+
+        for (size_t i = 0; i < n; i++, bitmap_reader.Next()) {
+          p_vec[i] = bitmap_reader.IsNotSet()
+                         ? NA_REAL
+                         : std::stod(decimals_arr.FormatValue(i).c_str());
+        }
+      } else {
+        for (size_t i = 0; i < n; i++) {
+          p_vec[i] = std::stod(decimals_arr.FormatValue(i).c_str());
+        }
+      }
+    }
+    return Status::OK();
+  }
+};
+
+struct Converter_NotHandled {
+  static SEXP Allocate(R_xlen_t n, const ArrayVector& arrays) {
+    stop(tfm::format("cannot handle Array of type %s", arrays[0]->type()->name()));
+    return R_NilValue;
+  }
+
+  static Status Ingest(SEXP data_, const std::shared_ptr<arrow::Array>& array,
+                       R_xlen_t start, R_xlen_t n) {
+    return Status::RError("Not handled");
+  }
+};
+
+// Most converter can ingest in parallel
+template <typename Converter>
+constexpr bool parallel_ingest() {
```

Comment:
Why don't you make this a static member of `Converter` classes?

Reply:
✅